### PR TITLE
Update and lint repository

### PR DESCRIPTION
**Pull request by tebeco**
The PR does absolutly 0 code code, It just 
* add artifacts/files to the `sln`
* add `.editorconfig`
  **please check the rules are ok**
* fix all warning / error from Roslyn now that there's an `.editorconfig`
* add a `global.json` to have a minimal SDK version for `dotnet tool`
* add a `tool-manifest` to version different dotnet tool used in this repo and version them
* add `dotnet-format` as a local tool
* run `dotnet format` (run `dotnet tool restore` before to restore all local tools for the tool manifest)

I did not yet add `dotnet tool restore` and `dotnet format --check` to the CI as it would first need a small discussion about that.
usually `dotnet format --check` is not in the build script itself, but rather a "check" aside the build itself, it take ~20 sec to run as it just parse/lint
It's just supposed to crash the status of the build to force the user to lint his code properly before pushing.
Linting code as part of the CI build does not make much sense to be honest.

What could be done but is completely overkill :
It would require to => lint, has change ?
no : build
yes : commit / push / abort this build
=> the push will re-trigger the build => lint => no change

So what i did :
* just version the manifest file
* never use it as part of the main CI
* if you want to add an additional `Pipeline/Check` for PR, with `dotnet restore tool` and `dotnet format --check` that would probably be good to update the README or error message to tell contributor that they do have to run `dotnet format` before pushing

Side note :
Due to weird reason this repo does generate a diff on some files, but `git status` is empty.
My guess is that it's a CRLF issue because the repo does not have any `.gitattributes` so it would need to have something like :
```
* text=auto
```
then run `git add --renormalize .`
(which is the commit named `git add --renormalize .` in this PR)
For details, see NDC video from @ethomson NDC video : https://youtu.be/VxmL6g9uy-A
from 0:00 to ~30:00  
the TLDR is here : https://youtu.be/VxmL6g9uy-A?t=1559

### CODE REVIEW

**Comment by tebeco** [comment]
This PR does not intend to change any code logic or whatever, it's just a kind of "cleanup" and adding repository manifests like `global.json`, `editorconfig` set of rules, `WarnAsError`, `C# 8`, `gitattributes` and all the potential associated fixes

**Comment by tebeco** [comment]
@LeeCampbell 

Also, the big issue for me is that i don't have `netcoreapp1.x` installed anywhere every (End Of Life is already reach way long time ago)
* the next PR will just bump to `netstandard2.0` the lib, and the tests/samples to  `netcoreapp2.1`(LTS), do I add `netcoreapp2.2` (EOL is passed) ?
* then next PR will try to add `netstandard2.1` / `netcoreapp3.1`, as the API used in this lib are not compliant and blow up in the test right now in `3.1` (probably a change in Synchronous IO that triggers a `Dispose` to the TextReader => way to big change for both other PR

**Comment by LeeCampbell** [comment]
Sorry to be a spoilt sport, but 18 commit and 100+ file changes!
General PR etiquette is to have small focused PRs (e.g. https://gist.github.com/mikepea/863f63d6e37281e329f8, https://medium.com/better-programming/pull-request-etiquettes-for-reviewer-and-author-f4e80360f92c, etc)

**Comment by tebeco** [comment]
so my question was ... what commit do you want to keep

**Comment by tebeco** [comment]
as the original pr contained too much i originally asked how to split it 
got no clue so i kinda go blind

can you please be more explicit on what part is interesting 
or if you are ok with 20 PR
